enterprise/cypher/physical-planning/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotConfigurationTest.scala

```diff
@@ -0,0 +1,152 @@
+
+    // when
+    slots.newReference("x", nullable = false, CTNumber)
+
+    // then
+    slots("x") should equal(RefSlot(0, false, CTNumber))
+  }
+
+  test("allocating same variable name with compatible types should work and get the upper bound type 2") {
+    // given
+    val slots = SlotConfiguration.empty
+    slots.newReference("x", nullable = false, CTAny)
+
+    // when
+    slots.newReference("x", nullable = false, CTNumber)
+
+    // then
+    slots("x") should equal(RefSlot(0, false, CTAny))
+  }
+
+  test("allocating same variable name with compatible types should work and get the upper bound type 3") {
+    // given
+    val slots = SlotConfiguration.empty
+    slots.newReference("x", nullable = true, CTMap)
+
+    // when
+    slots.newReference("x", nullable = false, CTAny)
+
+    // then
+    slots("x") should equal(RefSlot(0, true, CTAny))
+  }
+
+  test("allocating same variable name with compatible types should work and get the upper bound type 4") {
+    // given
+    val slots = SlotConfiguration.empty
+    slots.newReference("x", nullable = false, CTList(CTNumber))
+
+    // when
+    slots.newReference("x", nullable = true, CTList(CTInteger))
+
+    // then
+    slots("x") should equal(RefSlot(0, true, CTList(CTNumber)))
+  }
+
+  test("can't overwrite variable name by mistake1") {
+    // given
+    val slots = SlotConfiguration.empty
+    slots.newLong("x", nullable = false, CTNode)
+
+    // when && then
+    intercept[InternalException](slots.newLong("x", nullable = false, CTRelationship))
+  }
+
+  test("can't overwrite variable name by mistake2") {
+    // given
+    val slots = SlotConfiguration.empty
+    slots.newLong("x", nullable = false, CTNode)
+
+    // when && then
+    intercept[InternalException](slots.newReference("x", nullable = false, CTNode))
+  }
+
+  test("can't overwrite variable name by mistake3") {
+    // given
+    val slots = SlotConfiguration.empty
+    slots.newReference("x", nullable = false, CTNode)
+
+    // when && then
+    intercept[InternalException](slots.newLong("x", nullable = false, CTNode))
+  }
+
+  test("can't overwrite variable name by mistake4") {
+    // given
+    val slots = SlotConfiguration.empty
+    slots.newReference("x", nullable = false, CTNode)
+
+    // when && then
+    intercept[InternalException](slots.newReference("x", nullable = false, CTRelationship))
+  }
+
+  test("deepClone creates an immutable copy") {
+    // given
+    val slots = SlotConfiguration(Map(
+      "x" -> LongSlot(0, nullable = false, CTNode),
+      "y" -> LongSlot(1, nullable = false, CTNode)),
+      numberOfLongs = 2, numberOfReferences = 0)
+    val clone: SlotConfiguration = slots.copy()
+    slots should equal(clone)
+
+    // when
+    slots.newReference("a", nullable = false, CTNode)
+
+    // then
+    slots("a") should equal(RefSlot(0, nullable = false, CTNode))
+    clone.get("a") shouldBe empty
+    clone.numberOfReferences should equal(0)
```

Comment:
I really like this test. It would be even better though if it had some more assertions of that the longs are preserved, and that aliases are still the same.

Reply:
I just moved this test, but sure, I'll add some more assertions

Reply:
nice!